pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,432 @@
+		interest_rate: &InterestRate<Rate>,
+	) -> Result<Vec<CashflowPayment<Balance>>, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureDiv,
+		Rate: FixedPointNumber + EnsureDiv,
+	{
+		let start_date = date::from_seconds(origination_date)?;
+		let end_date = date::from_seconds(self.maturity.date())?;
+
+		let (timeflow, periods_per_year) = match &self.interest_payments {
+			InterestPayments::None => (vec![], 1),
+			InterestPayments::Monthly(reference_day) => (
+				date::monthly_intervals(start_date, end_date, (*reference_day).into())?,
+				12,
+			),
+		};
+
+		let total_weight = timeflow
+			.iter()
+			.map(|(_, weight)| weight)
+			.try_fold(0, |a, b| a.ensure_add(*b))?;
+
+		let interest_per_period = interest_rate
+			.per_year()
+			.ensure_div(Rate::saturating_from_integer(periods_per_year))?;
+
+		timeflow
+			.into_iter()
+			.map(|(date, weight)| {
+				let proportion = FixedU128::ensure_from_rational(weight, total_weight)?;
+				let principal = proportion.ensure_mul_int(principal)?;
+				let interest = interest_per_period.ensure_mul_int(principal)?;
+
+				Ok(CashflowPayment {
+					when: date::into_seconds(date)?,
+					principal,
+					interest,
+				})
+			})
+			.collect()
+	}
+
+	pub fn expected_payment<Balance, Rate>(
+		&self,
+		origination_date: Seconds,
+		principal: Balance,
+		interest_rate: &InterestRate<Rate>,
+		until: Seconds,
+	) -> Result<Balance, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureDiv,
+		Rate: FixedPointNumber,
+	{
+		let cashflow = self.generate_cashflows(origination_date, principal, interest_rate)?;
+
+		let total_amount = cashflow
+			.into_iter()
+			.take_while(|payment| payment.when < until)
+			.map(|payment| payment.principal.ensure_add(payment.interest))
+			.try_fold(Balance::zero(), |a, b| a.ensure_add(b?))?;
+
+		Ok(total_amount)
+	}
+}
+
+mod date {
+	use super::*;
+
+	pub fn from_seconds(date_in_seconds: Seconds) -> Result<NaiveDate, DispatchError> {
+		Ok(DateTime::from_timestamp(date_in_seconds.ensure_into()?, 0)
+			.ok_or("Invalid date in seconds, qed")?
+			.date_naive())
+	}
+
+	pub fn into_seconds(date: NaiveDate) -> Result<Seconds, DispatchError> {
+		Ok(date
+			.and_hms_opt(23, 59, 59) // Until the last second on the day
+			.ok_or("Invalid h/m/s, qed")?
+			.and_utc()
+			.timestamp()
+			.ensure_into()?)
+	}
+
+	pub fn next_month_with_day(date: NaiveDate, day: u32) -> Option<NaiveDate> {
+		let (month, year) = match date.month() {
+			12 => (1, date.year() + 1),
+			n => (n + 1, date.year()),
+		};
+
+		NaiveDate::from_ymd_opt(year, month, day)
+	}
+
+	pub fn monthly(
+		start_date: NaiveDate,
+		end_date: NaiveDate,
+		reference_day: u32,
+	) -> Result<Vec<NaiveDate>, DispatchError> {
+		if start_date >= end_date {
+			return Err(DispatchError::Other("Cashflow must start before it ends"));
+		}
+
+		if reference_day != DAY_1 {
+			return Err(DispatchError::Other(
+				"Only day 1 as reference is supported by now",
+			));
+		}
```

Comment:
I cannot return custom errors without access to `T: Config` from this. I'm not sure if we're okay with the `Other` variant for these errors.